src/source/src/table.rs

```diff
@@ -89,34 +88,42 @@ impl TableDmlHandle {
             // tasks to the compute nodes, so this'll be temporarily unavailable, so we throw an
             // error instead of asserting here.
             // TODO: may reject DML when streaming executors are not recovered.
-            let tx = self
-                .core
-                .read()
-                .changes_txs
-                .iter()
-                .choose(&mut rand::thread_rng())
-                .context("no available table reader in streaming source executors")?
-                .clone();
+            let tx = {
+                let guard = self.core.read();
+                if guard.changes_txs.is_empty() {
+                    return Err(RwError::from(anyhow!(
+                        "no available table reader in streaming source executors"
+                    )));
+                }
+                let len = guard.changes_txs.len();
+                guard
+                    .changes_txs
+                    .get((txn_msg.txn_id() % len as u64) as usize)
```

Comment:
I use `txn_msg.txn_id() % len` to choose the channel, so that data from the same transaction should be sent to the same `DMLExecutor`. Am I right?

Reply:
Yes, I think the same `changes_txs` can forward the same txn into the same dml executor.